docs/source/python/extending_types.rst

```diff
@@ -0,0 +1,115 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. currentmodule:: pyarrow
+.. _extending_types:
+
+Extending pyarrow
+=================
+
+Defining extension types ("user-defined types")
+-----------------------------------------------
+
+Arrow has the notion of extension types in the metadata specification as a
+possiblity to extend the built-in types. This is done by annotating any of the
+built-in Arrow logical types (the "storage type") with a custom type name and
+optional serialized representation ("ARROW:extension:name" and
+"ARROW:extension:metadata" keys in the Field’s custom_metadata of an IPC
+message).
+See the :ref:`format_metadata_extension_types` section of the metadata
+specification for more details.
+
+Pyarrow allows you to define such extension types from Python.
+
+There are currently two ways:
+
+* Subclassing :class:`ExtensionType`: the (de)serialization is based on pickle.
+  This is a good option for an extension type that is only used from Python.
+* Subclassing :class:`GenericExtensionType`: this allows to give a custom
+  Python-independent name and serialized metadata, that can potentially be
+  recognized by other (non-Python) Arrow implementations such as PySpark.
+
+
+For example, we could define a custom UUID type for 128-bit numbers which can
+be represented as ``FixedSizeBinary`` type with 16 bytes.
+Using the first approach, we create a `UuidType` subclass, and implement the
+``__reduce__`` dunder method to ensure the class can be properly pickled::
+
+    class UuidType(pa.ExtensionType):
+
+        def __init__(self):
+            pa.ExtensionType.__init__(self, pa.binary(16))
+
+        def __reduce__(self):
+            return UuidType, ()
+
+This can now be used to create arrays and tables holding the extension type::
+
+    >>> uuid_type = UuidType()
+    >>> uuid_type.extension_name
```

Comment:
Should `extension_name` be a private attrribute so the user doesn't *have* to have it as part of their classes' public api?
i.e. `s/extension_name/_extension_name` and if the user wants to opt in they can implement a property:
```python
@property
def extension_name(self):
    return self._extension_name
```
Similarly for `storage_type`.

Alternatively perhaps they can be "namespaced" - e.g.
```python
uuid_type.arrow.extension_name
uuid_type.arrow.storage_type
```

Reply:
@dhirschfeld is there a reason you wouldn't want it publicly? 
This is currently matching the C++ API where the extension name and storage type are methods on the type class. Those two properties are also quite fundamental to the concept of "extension types" in arrow, so I would personally not hide them.

Reply:
It prevents me having my own `storage_type` attribute which means something different in the context of my library.

...I don't actually have a `storage_type` attribute so it's just a hypothetical argument and thinking about it some more you could namespace it yourself:
```python
    @property
    def arrow_storage_type(self):
        return super().storage_type
```

Reply:
The principle is that I don't necessarily want my users to know the classes they're instantiating are being serialized by arrow - that's just an implementation detail similar to how the pickle protocol is defined using a private `__reduce__` method rather than a public `pickle` method

i.e. implementing the pickle protocol doesn't pollute the public api of your class with random attributes or methods.

> *Those two properties are also quite fundamental to the concept of "extension types" in arrow*

They're fundamental to arrow but irrelevant to my users. My users just care that they can save their instances so they can be recreated later and for me, arrow will hopefully allow the recreation to happen in Python, R or C# - i.e. the benefit over just using pickle is potentially enable cross-language interop.